test/utils/Checkpoints.t.sol

```diff
@@ -0,0 +1,276 @@
+            }
+        }
+
+        // check lookup
+        assertTrue(_ckpts.lowerLookup(lookup) == lower);
+        assertTrue(_ckpts.upperLookup(lookup) == upper);
+        assertTrue(_ckpts.upperLookupRecent(lookup) == upper);
+    }
+}
+
+contract CheckpointsTrace160Test is Test {
+    using Checkpoints for Checkpoints.Trace160;
+
+    Checkpoints.Trace160 internal _ckpts;
+
+    // helpers
+    function _boundUint96(uint96 x, uint96 min, uint96 max) internal view returns (uint96) {
+        return SafeCast.toUint96(bound(uint256(x), uint256(min), uint256(max)));
+    }
+
+    function _prepareKeys(uint96[] memory keys, uint96 maxSpread) internal view {
+        uint96 lastKey = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = _boundUint96(keys[i], lastKey, lastKey + maxSpread);
+            keys[i] = key;
+            lastKey = key;
+        }
+    }
+
+    function _assertLatestCheckpoint(bool exist, uint96 key, uint160 value) internal {
+        (bool _exist, uint96 _key, uint160 _value) = _ckpts.latestCheckpoint();
+        assertTrue(_exist == exist);
+        assertTrue(_key == key);
+        assertTrue(_value == value);
+    }
+
+    // tests
+    function testPush(uint96[] memory keys, uint160[] memory values) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        // initial state
+        assertTrue(_ckpts.length() == 0);
+        assertTrue(_ckpts.latest() == 0);
+        _assertLatestCheckpoint(false, 0, 0);
+
+        uint256 duplicates = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = keys[i];
+            uint160 value = values[i % values.length];
+            if (i > 0 && key == keys[i - 1]) ++duplicates;
+
+            // push
+            _ckpts.push(key, value);
+
+            // check length & latest
+            assertTrue(_ckpts.length() == i + 1 - duplicates);
+            assertTrue(_ckpts.latest() == value);
+            _assertLatestCheckpoint(true, key, value);
+        }
+    }
+
+    function testLookup(uint96[] memory keys, uint160[] memory values, uint96 lookup) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        uint96 lastKey = keys.length == 0 ? 0 : keys[keys.length - 1];
+        lookup = _boundUint96(lookup, 0, lastKey + 64);
+
+        uint160 upper = 0;
+        uint160 lower = 0;
+        uint96 lowerKey = type(uint96).max;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = keys[i];
+            uint160 value = values[i % values.length];
+
+            // push
+            _ckpts.push(key, value);
+
+            // track expected result of lookups
+            if (key <= lookup) {
+                upper = value;
+            }
+            if (key >= lookup && (i == 0 || keys[i - 1] < lookup)) {
+                lowerKey = key;
+            }
+            if (key == lowerKey) {
+                lower = value;
+            }
```

Comment:
```suggestion
            if (key >= lookup && (i == 0 || keys[i - 1] < lookup)) {
                lowerKey = key;
                lower = value;
            }
```

Reply:
No that would not work. 
```solidity
if (key >= lookup && (i == 0 || keys[i - 1] < lookup)) {
    lowerKey = key;
}
```
is here to identify the key of the "lower lookup" checkpoint, and  
```
if (key == lowerKey) {
    lower = value;
}
```
is here to get the last value pushed to that checkpoint (in case of duplicated key)

Reply:
Yes, this is not obvious from the code, but the separate `if` is needed to process duplicate keys.